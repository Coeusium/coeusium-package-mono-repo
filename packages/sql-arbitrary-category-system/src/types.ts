export type { ConnectionInfo } from '@almaclaine/mysql-utils';

export interface CategorySet {
  id: string;
  name: string;
}

export interface Category {
  set_id: string;
  id: string;
  name: string;
  parent?: string;
}
