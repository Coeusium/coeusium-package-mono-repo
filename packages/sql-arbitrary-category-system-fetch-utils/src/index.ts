export * from './sql-arbitrary-category-system-fetch-utils';
export type { Category, CategorySet } from './types';
