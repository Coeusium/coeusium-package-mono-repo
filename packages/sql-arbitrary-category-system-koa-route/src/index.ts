export * from './sql-arbitrary-category-system-koa-route';
export type { ConnectionInfo, Category, CategorySet } from './types';
