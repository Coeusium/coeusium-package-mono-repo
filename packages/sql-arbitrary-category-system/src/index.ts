export * from './sql-arbitrary-category-system';
export type { ConnectionInfo, Category, CategorySet } from './types';
