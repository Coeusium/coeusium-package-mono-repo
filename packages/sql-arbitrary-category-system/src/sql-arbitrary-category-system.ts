import {
  ConnectionInfo,
  deleteFromTableById,
  execute,
  getFromTableById,
  idExistsInTable,
  listFromTable,
  setupDatabase,
  readSQLFiles,
} from '@almaclaine/mysql-utils';
import { Category, CategorySet } from './types';
import { makeId } from '@almaclaine/general-utils';

export async function setupCategorySystem(dbInfo: ConnectionInfo) {
  await setupDatabase(dbInfo, 'sql_category_system', await readSQLFiles());
}

export async function destroy() {
  execute.destroyConnections();
}

// Category Set Utilities

export async function categorySetIdExists(dbInfo: ConnectionInfo, id: string) {
  return await idExistsInTable(dbInfo, 'category_set', id);
}

export async function addCategorySet(dbInfo: ConnectionInfo, setName: string) {
  const sql = `INSERT INTO category_set (id, name) VALUES (?, ?);`;
  let id = makeId();
  while (await categorySetIdExists(dbInfo, id)) id = makeId();
  await execute(dbInfo, sql, [id, setName]);
  return id;
}

export async function getCategorySet(dbInfo: ConnectionInfo, id: string) {
  return await getFromTableById<CategorySet>(dbInfo, 'category_set', id);
}

export async function listCategorySets(
  dbInfo: ConnectionInfo,
  page = 0,
  limit = 20,
) {
  return await listFromTable<CategorySet>(dbInfo, 'category_set', page, limit);
}

export async function deleteCategorySet(dbInfo: ConnectionInfo, id: string) {
  await deleteFromTableById(dbInfo, 'category_set', id);
}

// Category Utilities

export async function categoryIdExists(dbInfo: ConnectionInfo, id: string) {
  return await idExistsInTable(dbInfo, 'category', id);
}

export async function addCategory(
  dbInfo: ConnectionInfo,
  categoryName: string,
  setId: string,
  parentId?: string,
) {
  const sql = `INSERT INTO category (set_id, id, name) VALUES (?, ?, ?);`;
  let id = makeId();
  while (await categoryIdExists(dbInfo, id)) id = makeId();
  await execute(dbInfo, sql, [setId, id, categoryName]);
  if (parentId && (await categoryIdExists(dbInfo, parentId))) {
    const sql2 = `UPDATE category SET parent = ? WHERE id = ?;`;
    await execute(dbInfo, sql2, [parentId, id]);
  }
  return id;
}

export async function getCategory(dbInfo: ConnectionInfo, id: string) {
  return await getFromTableById<Category>(dbInfo, 'category', id);
}

export async function getTopLevelCategories(
  dbInfo: ConnectionInfo,
  setId: string,
) {
  const sql = `SELECT * FROM category WHERE set_id = ? AND parent IS NULL;`;
  return ((await execute(dbInfo, sql, [setId])) as Category[]) || [];
}

export async function getCategoryChildren(dbInfo: ConnectionInfo, id: string) {
  const sql = `SELECT * FROM category WHERE parent = ?;`;
  return ((await execute(dbInfo, sql, [id])) as Category[]) || [];
}

export async function getCategoryDescendants(
  dbInfo: ConnectionInfo,
  id: string,
) {
  const children = await getCategoryChildren(dbInfo, id);
  const rest = (
    await Promise.all(
      children.map(
        async ({ id: childId }) =>
          await getCategoryDescendants(dbInfo, childId),
      ),
    )
  ).reduce((a, c) => a.concat(c), []);
  return [...children, ...rest];
}

export async function deleteCategory(dbInfo: ConnectionInfo, id: string) {
  await deleteFromTableById(dbInfo, 'category', id);
}
