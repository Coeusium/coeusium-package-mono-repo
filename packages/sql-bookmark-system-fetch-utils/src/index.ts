export * from './sql-bookmark-system-fetch-utils';
export type { Bookmark } from './types';
