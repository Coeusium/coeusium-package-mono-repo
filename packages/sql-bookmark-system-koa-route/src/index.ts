export * from './sql-bookmark-system-koa-route';
export type { ConnectionInfo, Bookmark } from './types';
