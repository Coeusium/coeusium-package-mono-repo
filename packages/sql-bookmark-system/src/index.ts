export * from './sql-bookmark-system';
export type { ConnectionInfo, Bookmark } from './types';
