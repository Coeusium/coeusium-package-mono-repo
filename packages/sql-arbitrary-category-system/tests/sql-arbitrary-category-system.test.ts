import * as sqlCategorySystem from '../src';

describe('sql-category-system', () => {
  it('Has Correct API', () => {
    const keys = Object.keys(sqlCategorySystem);
    expect(keys.length).toBe(14);
    expect(keys).toMatchSnapshot();
  });
});
