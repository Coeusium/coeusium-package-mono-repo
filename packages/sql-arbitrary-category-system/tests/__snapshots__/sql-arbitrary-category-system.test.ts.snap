// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql-category-system Has Correct API 1`] = `
Array [
  "setupCategorySystem",
  "destroy",
  "categorySetIdExists",
  "addCategorySet",
  "getCategorySet",
  "listCategorySets",
  "deleteCategorySet",
  "categoryIdExists",
  "addCategory",
  "getCategory",
  "getTopLevelCategories",
  "getCategoryChildren",
  "getCategoryDescendants",
  "deleteCategory",
]
`;
